rfcs/0000_intelligent_resampling.md

```diff
@@ -0,0 +1,368 @@
+
+The key challenge is deciding whether to:
+- Learn an entirely new object model (when encountering a truly novel object)
+- Update an existing model; this might be necessary when encountering a known object in a new setting (e.g. different lighting), if the model is incomplete (e.g. unexplored part), or it has changed (e.g., your favorite mug now has a chip in it)
+- Simply eliminate incorrect hypotheses (when the models are sufficient but we're at the wrong location/on a different object)
+
+To treat learning and inference as lying on a continuum, we can utilize metadata stored in object models as a proxy/heuristic to use prediction errors in one way or another. Heuristics may be:
+
+- **Hypothesis coverage**: If some hypotheses remain valid after prediction errors, the existing models are likely sufficient and we should focus on hypothesis elimination. However, if all hypotheses are eliminated (or less than a certain percentage threshold), we may need to learn a new model or update an existing one.
+
+- **Observation frequency** (e.g., `_observation_count` in `GridObjectModel`): Frequently visited locations with high observation counts suggest the model is well-learned at that location, biasing toward hypothesis elimination rather than model updates.
+
+- **Compositionality**: When prediction errors occur at specific locations on an otherwise well-matched familiar object, consider learning the modification as a compositional element rather than updating the base model. This preserves the integrity of well-learned models while capturing local variations or modifications.
+
+- **Error magnitude thresholds**: Very large prediction errors across all features may suggest a novel object requiring a new model. Moderate errors might indicate the need for model updates or hypothesis refinement.
+
+- **Feature-specific patterns**: If morphological features match but non-morphological features (like color) differ significantly, this might indicate object variations / keyframe / need to separate morphological and feature models.
+
+#### Example of Chipped Coffee Mug 
+
+Consider the TBP mug that we are very familiar that is now slightly chipped. Below are how the above heuristics may apply:
+
+1. **Initial detection**: When sensing the chipped area, we would get prediction errors since the surface geometry differs from our stored model.
+2. **Hypothesis coverage**: The hypothesis for where we are on the mug would remain valid across many locations during movement (handle, sides, rim); it would only result in a prediction error / risk elimination when we move to the chip, the point at which we would want to signal learning in a model.
+3. **Observation frequency**: The mug model has high observation counts from frequent use, suggesting we shouldn't discard the entire model.
+4. **Feature patterns**: Morphological features (surface curvature) would differ at the chip location, but non-morphological features (color, texture) might remain consistent.
```

Comment:
Why would color and texture stay the same? If it is chipped then those would be gone too, right?